internal/pkg/runtime/engine/apptainer/process_linux.go

```diff
@@ -77,8 +77,18 @@ func (e *EngineOperations) StartProcess(masterConnFd int) error {
 	bootInstance := isInstance && e.EngineConfig.GetBootInstance()
 	shimProcess := false
 
+	customCwd := true
+	switch e.EngineConfig.OciConfig.Process.Cwd {
+	case e.EngineConfig.GetCwd(), e.EngineConfig.GetHomeDest():
+		customCwd = false
+	}
+
 	if err := os.Chdir(e.EngineConfig.OciConfig.Process.Cwd); err != nil {
+		if customCwd {
+			return fmt.Errorf("failed to set working directory: %s", err)
+		}
 		if err := os.Chdir(e.EngineConfig.GetHomeDest()); err != nil {
+			sylog.Warningf("Error setting the working directory. Using '/' instead: %s", err)
```

Comment:
There shouldn't be a warning here, because this is normal operation in the case of `--contain`.  A debug message could be helpful.